packages/styletron-inferno/src/styled.js

```diff
@@ -0,0 +1,124 @@
+const createElement = require('inferno-create-element');
+const {injectStylePrefixed} = require('styletron-utils');
+const {
+  assign,
+  isNil,
+  isObject,
+  isString,
+  isFunction
+} = require('./utils');
+
+const STYLETRON_KEY = '__STYLETRON';
+
+module.exports = styled;
+
+/**
+ * Helper function to create styled components
+ * @packagename styletron-inferno
+ * @param  {string|function} name   Tag name or component function/class
+ * @param  {function|object} styles Style object or function that returns a style object
+ * @return {function}               Styled component
+ *
+ * @example
+ * import { styled } from 'styletron-inferno';
+ *
+ * const Panel = styled('div', {
+ *   backgroundColor: 'lightblue',
+ *   fontSize: '12px'
+ * });
+ *
+ * <Panel>Hello World</Panel>
+ *
+ * @example
+ * import { styled } from 'styletron-inferno';
+ *
+ * const Panel = styled('div', (props) => ({
+ *   backgroundColor: props.alert ? 'orange' : 'lightblue',
+ *   fontSize: '12px'
+ * }));
+ *
+ * <Panel alert>Danger!</Panel>
+ *
+ * @example
+ * import { styled } from 'styletron-inferno';
+ *
+ * const DeluxePanel = styled(Panel, (props) => ({
+ *   backgroundColor: props.alert ? 'red' : 'lime',
+ *   boxShadow: '3px 3px 3px darkgray',
+ *   color: 'white'
+ * }));
+ *
+ * <DeluxePanel>Bonjour Monde</DeluxePanel>
+ */
+function styled(name, styles) {
+
+  // Styled component
+  if (name && name.hasOwnProperty(STYLETRON_KEY)) {
+    const component = name[STYLETRON_KEY];
+    const stylesArray = component.stylesArray.concat(styles);
+    return createStyledComponent(component.name, stylesArray);
+
+  // Tag name or non-styled component
+  } else if (isString(name) || isFunction(name)) {
+    return createStyledComponent(name, [styles]);
+  }
+
+  throw new Error('Must pass a tag name or component');
+}
+
+function createStyledComponent(name, stylesArray) {
+
+  function StyledComponent(props, context) {
+    const newProps = assign({}, props);
+    const styles = resolveStyles(stylesArray, props, context);
+    const className = injectStylePrefixed(context.styletron, styles);
+    newProps.className = mergeClassNames(props.className, className);
+    if (isFunction(props.innerRef)) {
+      newProps.ref = props.innerRef;
+      delete newProps.innerRef;
+    }
+    return createElement(name, newProps);
+  }
+
+  StyledComponent[STYLETRON_KEY] = {name, stylesArray};
+
+  return StyledComponent;
+}
+
+function resolveStyles(stylesArray, props, context) {
+  const resolvedStyles = {};
+  for (let i = 0, l = stylesArray.length, styles; i < l; i++) {
+    styles = stylesArray[i];
+    if (!isNil(styles)) {
+      if (isFunction(styles)) {
+        assign(resolvedStyles, styles(props, context));
+      } else if (isObject(styles)) {
+        assign(resolvedStyles, styles);
+      }
+    }
+  }
+  return resolvedStyles;
+}
+
+const SPACE_REGEX = /\s+/;
+
+function mergeClassNames(classNameA, classNameB) {
```

Comment:
I think the addition of de-duplication of class names is a really nice feature, but it potentially makes this function somewhat expensive. I suspect there's room for further optimization here.

Two ideas:
1.   Use a stateful regex iteration within a while loop instead of using `String.prototype.split` which involves an array allocation. I suspect building the set of unique classes should be faster this way.

2. Use of a modified version of `injectStyle` from `styletron-utils` that returns an object with atomic class keys. Then merging can simply be repeated `Object.assign` and then an iteration of the final object keys to build the final class name string. I hadn't thought of this before, but this would definitely have an impact on https://github.com/rtsao/styletron/pull/91